File: xml-store/src/main/java/com/jarredweb/editor/app/NashJs.java
package com.jarredweb.editor.app;

import com.coveo.nashorn_modules.FilesystemFolder;
import com.coveo.nashorn_modules.Require;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.util.Arrays;
import javax.script.Invocable;
import javax.script.ScriptContext;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;
import jdk.nashorn.api.scripting.NashornScriptEngine;
import jdk.nashorn.api.scripting.ScriptObjectMirror;

public class NashJs {

    private static final NashornScriptEngine ENGINE = (NashornScriptEngine) new ScriptEngineManager().getEngineByName("nashorn");

    public static void main(String... args) {
        try {
            FilesystemFolder rootFolder = FilesystemFolder.create(new File("www"), "UTF-8");
            Require.enable(ENGINE, rootFolder);
            ScriptContext context = ENGINE.getContext();
            context.setAttribute("runner", new NashJs(), ScriptContext.ENGINE_SCOPE);

            printHello();
            loadKnockout();
            loadNashjs();
            loadVuejs();
        } catch (ScriptException | FileNotFoundException | NoSuchMethodException ex) {
            ex.printStackTrace(System.err);
        }
    }

    public static void printHello() throws ScriptException {
        ENGINE.eval("print('Hello from javascript!');");
    }

    public static void loadNashjs() throws ScriptException, FileNotFoundException, NoSuchMethodException {
        ENGINE.eval(new FileReader("www/catalog/nashjs.js"));
        Invocable invocable = (Invocable) ENGINE;

        Object result = invocable.invokeFunction("sayHello", "John Doe");
        System.out.println(result);
        System.out.println(result.getClass());
    }

    public static void loadKnockout() throws ScriptException, FileNotFoundException, NoSuchMethodException {
        ENGINE.eval(new FileReader("www/public/js/knockout-3.4.2.js"));
        Invocable invocable = (Invocable) ENGINE;
    }

    public static void loadVuejs() throws ScriptException, FileNotFoundException, NoSuchMethodException {
        ENGINE.eval(new FileReader("www/public/js/vue.js"));
        Invocable invocable = (Invocable) ENGINE;

        Object result = invocable.invokeFunction("printVueVersion");
        System.out.println(result);

    }

    public static void printType(Object object) {
        System.out.println(object.getClass());
    }

    public static void printObjectMirror(ScriptObjectMirror mirror) {
        System.out.println(mirror.getClassName() + ": " + Arrays.toString(mirror.getOwnKeys(true)));
    }

    public static String javaHello(String name) {
        return String.format("Hello %s from Java!", name);
    }

    public static void getFullName(ScriptObjectMirror person) {
        System.out.println("Full name is: " + person.callMember("getFullName"));
    }

    public int add(int a, int b) {
        return a + b;
    }
}

File: xml-store/www/catalog/nashjs.js
print(java.lang.System.currentTimeMillis());

var file = new java.io.File("browser.min.js");
print(file.getAbsolutePath());
print(file.absolutePath);

var sayHello = function(name) {
    print('Hello, ' + name + '!');
    return 'hello from javascript';
};

var NashJs = Java.type('com.jarredweb.editor.app.NashJs');

// call the static method
var greetingResult = NashJs.javaHello('Mars');
print(greetingResult);

// create a new intance of MyJavaClass
var runjs = new NashJs();
var calcResult = runjs.add(1, 2);
print(calcResult);

// check out java types
NashJs.printType('Hello');
// class java.lang.String
NashJs.printType(123);
// class java.lang.Integer
NashJs.printType(12.34);
// class java.lang.Double
NashJs.printType(true);
// class java.lang.Boolean

NashJs.printType(new Number(123));
// class jdk.nashorn.internal.objects.NativeNumber
// class jdk.nashorn.api.scripting.ScriptObjectMirror
NashJs.printType(new Date());
// class jdk.nashorn.internal.objects.NativeDate
// class jdk.nashorn.api.scripting.ScriptObjectMirror
NashJs.printType(new RegExp());
// class jdk.nashorn.internal.objects.NativeRegExp
// class jdk.nashorn.api.scripting.ScriptObjectMirror
NashJs.printType({ foo: 'bar' });
// class jdk.nashorn.internal.scripts.J04
// class jdk.nashorn.api.scripting.ScriptObjectMirror

NashJs.printObjectMirror(new Number(123));
//Number: []
NashJs.printObjectMirror(new Date());
//Date: []
NashJs.printObjectMirror(new RegExp());
//RegExp: [lastIndex, source, global, ignoreCase, multiline]
NashJs.printObjectMirror({
    foo: 'bar',
    bar: 'foo'
});
//Object: [foo, bar]

calcResult = runner.add(10, 20);
print('result is ' + calcResult);
//result is 30

function Person(firstName, lastName) {
    this.firstName = firstName;
    this.lastName = lastName;
    this.getFullName = function() {
        return this.firstName + ' ' + this.lastName;
    }
}

var person = new Person('John', 'Doe');
NashJs.getFullName(person);
// Full name is: John Doe

// var Vue = require('www/catalog/vue.js');

// var vue = Vue.component({
//     name: 'sample',
//     data: function() {
//         return {
//             version: 1.0
//         }
//     },
//     methods: {
//         hello: function() {
//             return "you got this!";
//         }
//     }
// });

// var printVueVersion = function() {
//     return vue.data.version;
// }


// var content = readFully('catalog.xml');
// print(content);

// var files = `ls -l`; // get file listing as a string
// var lines = files.split("\n");
// for (var l in lines) {
//     var line = lines[l];
//     print(line);
// }

// print($ENV["JAVA_HOME"])
// print($ENV["PATH"])
// print($ENV.JAVA_HOME)
// print($ENV.PATH)